platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,310 @@
+
+The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
+consensus and block signing.
+Roster, an immutable subset of the address book, will provide this mechanism, ensuring efficient and secure key
+management for TSS operations.
+This proposal attempts to provide a specification for the behavior of rosters starting from their creation from the
+Candidate Address Book (CAB) to their terminal state within the TSS specification.
+A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.
+
+The Future Address Book (FAB) is a dynamic list maintained within the App that reflects the desired future state of the
+network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
+nodes.
+The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
+book.
+
+This proposal specifies that Roster(s) will be created from the CAB by the Hedera app and passed to the platform code.
+The mechanism for doing so is detailed below.
+
+### Requirements
+
+- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use rosters instead.
+- Address book paradigm discontinued within the platform codebase, only use rosters instead
+
+### Architecture and/or Components
+
+###### Roster API
+
+The Hedera app is already responsible for managing the address book.
+We propose that it continues to do so, and at adoption time, create a candidate Roster object from the CAB and set it in
+the state.
+The State will contain one map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```.
+A Roster has a data structure as follows:
+
+```
+class Roster {
+    List~RosterEntry~ entries
+}
+
+class RosterEntry {
+    NodeId nodeId
+    long weight
+    X509Certificate signingCert
+    PairingPublicKey tssEcKey
+    List~ServiceEndpoint~ gossipEndpoints
+}
+
+class ServiceEndpoint {
+    bytes ipAddressV4
+    int32 port
+    string domain_name
+}
+
+Roster "1" *-- "many" RosterEntry
+RosterEntry "1" *-- "many" ServiceEndpoint
+
+```
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, the optional
+previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.
```

Comment:
Yes Ed, that's the idea ( off by 1 errors aside, which I'm personally not bothering to try to get straight... that's a job for the implementer :D )

Reply:
I don't think I've documented how the round association will work (earlier idea was to include it in the Roster itself but that proposal had been witdrawn) so this is a valuable idea.

You're proposing an additional data structure - a queue of Roster hashes - where the index of the queue represents round numbers. One thing i'd add is to consider making the round vs roster hash relationship explicit. It will be easy to introduce bugs if implicit—wrong offset, queue pop at the wrong time or by the wrong thread, etc.).

Do we see ever having to do a round lookup (e.g. what roster was used x rounds ago)? 

I'd love to include this once we work out these tiny details. Thanks for raising it!

Reply:
To be clear, the round-number-as-index structure isn't something I'm advocating for right now. Rather, it's a plan compatible with full DAB, which utilizes a queue. The proposal for the short-term solution to use a queue, so that fewer changes are required when the transition to full DAB *is* made. The section in my original comment called "Pre full DAB" describes the short term proposal.

> One thing i'd add is to consider making the round vs roster hash relationship explicit. It will be easy to introduce bugs if implicit—wrong offset, queue pop at the wrong time or by the wrong thread, etc.).

We need to be very stingy with what we put into the state, if we can implicitly store this info via index, we should. If we're introducing off-by-one bugs in this area of the code, we can all pack up and go home anyway.

> Do we see ever having to do a round lookup (e.g. what roster was used x rounds ago)?

Yes, this is how we will validate events in a fully DAB system.

Reply:
I'll setup a call to discuss this idea in detail.

Reply:
Thanks all for the discussion last week.

FYI I've introduced a new diagram capturing this idea, and presented the pre-DAB implementation in this proposal. The full-DAB implementation will be implemented as part of dynamic address book.